android/expoview/src/main/java/host/exp/exponent/ExpoUpdatesAppLoader.java

```diff
@@ -0,0 +1,237 @@
+import expo.modules.updates.loader.FileDownloader;
+import expo.modules.updates.loader.LoaderTask;
+import expo.modules.updates.manifest.Manifest;
+import host.exp.exponent.analytics.Analytics;
+import host.exp.exponent.analytics.EXL;
+import host.exp.exponent.di.NativeModuleDepsProvider;
+import host.exp.exponent.exceptions.ExceptionUtils;
+import host.exp.exponent.kernel.ExpoViewKernel;
+import host.exp.exponent.kernel.ExponentUrls;
+import host.exp.exponent.kernel.KernelConfig;
+import host.exp.exponent.storage.ExponentDB;
+import host.exp.exponent.storage.ExponentSharedPreferences;
+import host.exp.expoview.ExpoViewBuildConfig;
+import host.exp.expoview.Exponent;
+
+public abstract class ExpoUpdatesAppLoader {
+
+  @Inject
+  ExponentManifest mExponentManifest;
+
+  @Inject
+  ExponentSharedPreferences mExponentSharedPreferences;
+
+  @Inject
+  DatabaseHolder mDatabaseHolder;
+
+  private static final String TAG = ExpoUpdatesAppLoader.class.getSimpleName();
+
+  private String mManifestUrl;
+  private final boolean mUseCacheOnly;
+
+  private Context mContext;
+
+  public ExpoUpdatesAppLoader(String manifestUrl) {
+    this(manifestUrl, false, null);
+  }
+
+  public ExpoUpdatesAppLoader(String manifestUrl, boolean useCacheOnly, Context context) {
+    NativeModuleDepsProvider.getInstance().inject(ExpoUpdatesAppLoader.class, this);
+
+    mManifestUrl = manifestUrl;
+    mUseCacheOnly = useCacheOnly;
+    mContext = context;
+  }
+
+  public abstract void onOptimisticManifest(JSONObject optimisticManifest);
+
+  public abstract void onManifestCompleted(JSONObject manifest);
+
+  public abstract void onBundleCompleted(String localBundlePath);
+
+  public abstract void emitEvent(JSONObject params);
+
+  public abstract void onError(Exception e);
+
+  public abstract void onError(String e);
+
+  public void start() {
+    Uri manifestUrl = mExponentManifest.httpManifestUrl(mManifestUrl);
+
+    // TODO: handle development mode manifests
+
+    HashMap<String, Object> configMap = new HashMap<>();
+    configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_UPDATE_URL_KEY, manifestUrl);
+    // TODO: if we want to use a scopeKey from the manifest here,
+    //  need to either keep track of URL -> scopeKey mapping separately
+    configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_SCOPE_KEY_KEY, mManifestUrl);
+    configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_SDK_VERSION_KEY, Constants.SDK_VERSIONS);
+    configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_HAS_EMBEDDED_UPDATE, false);
+    configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_ENABLED_KEY, Constants.ARE_REMOTE_UPDATES_ENABLED);
+    if (mUseCacheOnly) {
+      configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_CHECK_ON_LAUNCH_KEY, "NEVER");
+      configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_LAUNCH_WAIT_MS_KEY, 0);
+    } else {
+      // TODO: decide about default launch behavior for development client
```

Comment:
Loading a stale update is problematic while in development. I think we always want to load the latest update like a browser would unless the developer is intentionally testing the caching behavior.

Reply:
I agree, though I think there is a somewhat larger discussion to be had here and will write up a Notion doc about it. How would a developer indicate they want to test the caching behavior rather than loading the latest update synchronously? (I'm assuming you are talking here about loading production bundles in the development client, as I don't think there's any ambiguity about what the behavior should be for development bundles.)

Reply:
That might be a good question for the folks thinking about the dev client module. The dev menu already lets us configure several settings and overriding the caching behavior could be an addition option.